Make the auction example end-to-end (#6477)
IntersectMBO/plutus: doc/docusaurus/docs/auction-smart-contract/end-to-end/mint.md

---
sidebar_position: 15
---

# Minting the Token to be Auctioned

Before we can start the auction, we need to mint a token to be auctioned.

> On Cardano, each token has a currency symbol and a token name.

Reviewer: ```suggestion
On Cardano, each token is uniquely identified by its currency symbol and a token name.
```

Author: That doesn't sound accurate, since currency symbol + token name uniquely identify a class of tokens, but there can be more than one tokens of that class.

Reviewer: Right. The word "token" is overloaded. I meant "class of tokes". My goal with this change was to avoid new developers' common confusion about the currency symbol being enough to define/identify the token. Else, the original was fine. 

Being mindful of the context, does this sound better?
```suggestion
To do so, we first must determine the token's currency symbol and name.
```

To mint or burn tokens with a specific currency symbol (`currencySymbol`), a Plutus script whose hash matches `currencySymbol` must be provided, and is used to validate the minting or burning action.
Therefore, we'll first write the on-chain script, then compute its hash and use it as the currency symbol.

## On-chain Minting Policy Script

> The full minting policy code can be found at [AuctionMintingPolicy.hs](https://github.com/IntersectMBO/plutus-tx-template/blob/main/src/AuctionMintingPolicy.hs).

Reviewer 2: I noticed that minting policy uses Plutus V2. Do we have a task to update it to the Plutus V3?

Author: I think V2 is fine for this example, since there's no advantage (in fact there's disadvantage) to use Plutus V3. But if anyone wants to update the example to V3, you are welcome to.

The main logic is in the following function:

<LiteralInclude file="AuctionMintingPolicy.hs" language="haskell" title="AuctionMintingPolicy.hs" start="-- BLOCK1" end="-- BLOCK2" />

This script will pass if the following two conditions are met:

1. The transaction is signed by a specific public key.
2. The transaction mints exactly one token, whose currency symbol matches the script's hash (i.e., `ownCurrencySymbol ctx`).
  The token name can be anything.

> :pushpin: **NOTE**
>
> A token minted in this way is _not_ considered a non-fungible token (NFT) because, while only one token can be minted in a single transaction, multiple transactions can mint additional tokens with the same currency symbol and token name.
> To create a truly unique token, you would need a more complex minting policy, but for simplicity, that is not covered here.

## Compile and Generate Blueprint for the Minting Policy

Next, we need to compile the minting policy script and create its blueprint.
To do so, we first need to supply a public key hash, which the minting policy will use for checking condition 1 above.
Assuming the seller is the one minting the token, this should be the seller's public key hash.

> Open `GenMintingPolicyBlueprint.hs` in the `on-chain` directory, and replace `error "Replace with seller pkh"` with the content of `off-chain/seller.pkh`.

Reviewer 2: I couldn't find this file `GenMintingPolicyBlueprint.hs`

> Open `GenMintingPolicyBlueprint.hs` in the `on-chain` directory, and replace `error "Replace with seller pkh"` with the content of `off-chain/seller.pkh`.

Reviewer 2: Is there a reason to put auction validator script and minting policy script in two different blueprint files?
I can see how two different plutus scripts (validator, policy) together constitute a one logical "contract".

Author: To generate blueprint for the auction validator, we need to know the hash of the minting validator, and currently the best way to get the hash is from the blueprint.

The minting policy code comes with `plutus-tx-template`, so you can find it in the `on-chain` repository.
To compile it and generate the blueprint, navigate to the `on-chain` directory and run

```
cabal run gen-minting-policy-blueprint -- ../off-chain/plutus-auction-minting-policy.json
```

You may need to run `cabal update` before executing this command for the first time.

This should produce a blueprint file named `off-chain/plutus-auction-minting-policy.json`.

## Compile and Generate Blueprint for the Auction Validator

One final step before minting the token: since we want to lock the minted token at the script address corresponding to the auction validator,
we must supply the parameters (i.e., `AuctionParams`) to the auction validator, compile the auction validator, and calculate its script address.

> Open `AuctionValidator.hs` in the `on-chain` directory, and replace all placeholders:

Reviewer: ```suggestion
Open `GenAuctionValidatorBlueprint.hs` in the `on-chain` directory, and replace all placeholders:
```
